### PR TITLE
replace NotEmpty annotations with NotBlank

### LINKED ISSUE #508: Replace NotEmpty annotations with NotBlank
Replace `NotEmpty` annotations with `NotBlank` (or add the annotation for the fields that don't have a `NotEmpty` annotation) for the following fields:
- Author `firstName`
- Author `lastName`
- Book `title`
- Shelf `shelfName`
- Tag `name`

There should also be accompanying tests so that if an annotation were to be removed, we got informed by virtue of a failing test.

### PR DESCRIPTION
**Pull request by kateerfle**
## Summary of change

I have changed the NotEmpty annotations in the Author, Book, and Tag classes.  I did not do this for Shelf, although I did want to ask if I should skip CustomShelf only or both CustomShelf and Shelf? I am unsure of where to test these changes.  I originally thought in the AuthorTest.java and BookTest.java files, but now I am thinking it should be in the BookFormTest.java file?  I am also not sure if my tests are on the right track.  Some feedback would be greatly appreciated as I have never worked with these Java annotations, so I am not exactly sure how to test them.  Thank you!

## Related issue

Closes #508

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

For any of the optional checkboxes (e.g. the screenshots one), still check it if it does not apply.

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-i7lept44-rgJ9yB0A2vedJTLyyfkjKQ)

### CODE REVIEW

**Comment by knjk04** [comment]
> I have changed the NotEmpty annotations in the Author, Book, and Tag classes. I did not do this for Shelf, although I did want to ask if I should skip CustomShelf only or both CustomShelf and Shelf?

For both, please. Sorry, I should have said.

> I am unsure of where to test these changes. I originally thought in the AuthorTest.java and BookTest.java files, but now I am thinking it should be in the BookFormTest.java file? 

I think where you have them is good. We want the author-related tests to be in `AuthorTest` because it's testing something that's in the `Author` class. Same with the `BookTest` class.

> I am also not sure if my tests are on the right track. Some feedback would be greatly appreciated as I have never worked with these Java annotations, so I am not exactly sure how to test them. Thank you!

You're definitely on the right track. My approach would have been to test the right exception was thrown when we try to save the entity using the respective service class (e.g. save an `Author` object using `AuthorService`), but your way is much better as it avoids needing to use the services at all!

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 1]
```diff
@@ -114,4 +131,30 @@ void notRemoveNonOrphans() {
         // then
         assertThat(authorService.count()).isOne();
     }
+    
+    @Test
+    void notEmptyFirstName() {
+    	// given
+    	Author noFirst = new Author(" ", "Gaarder");
+    	
+    	validator = Validation.buildDefaultValidatorFactory().getValidator();
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:140]: Could you please move this up to the constructor so that we only do this once (also remove it out of the other test)?

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 2]
```diff
@@ -114,4 +131,30 @@ void notRemoveNonOrphans() {
...
+    	Set<ConstraintViolation<Author>> violations = validator.validateProperty(noFirst, "firstName");
+    	
+    	//  then
+    	assertThat(violations.size()).isEqualTo(1);
+    }
+    
+    @Test
+    void notEmptyLastName() {
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:148]: image

Every test class has the name of the class it's testing followed by 'should'. In `AuthorTest`, you'll see the display name is 'Author should'. Each test method follows on from that to complete the sentence (e.g. `Author should` ... `notRemoveNonOrphans`).

Could you please rename your methods to follow this style? For example, this method could be renamed to something like 'notAcceptEmptyLastName' (so that we get `Author should notAcceptEmptyLastName`)

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 3]
```diff
@@ -50,6 +61,12 @@ public void reset(@Autowired PredefinedShelfService predefinedShelfService) {
         resetBookService();
         toRead = predefinedShelfService.findToReadShelf();
     }
+    
+  //  @Before
+  //  public void setUp() {
+    //    ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
+      //  validator = factory.getValidator();
+   // }
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:69]: I don't think you'll need this any more, so this can be removed (and the commented out import)

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 4]
```diff
@@ -114,4 +131,30 @@ void notRemoveNonOrphans() {
         // then
         assertThat(authorService.count()).isOne();
     }
+    
+    @Test
+    void notEmptyFirstName() {
+    	// given
+    	Author noFirst = new Author(" ", "Gaarder");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:138]: For readability, could you please rename this to `authorWithEmptyFirstName`? To me, no first name would be null (absence of a value)? Also,

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 5]
```diff
@@ -114,4 +131,30 @@ void notRemoveNonOrphans() {
...
+    	Set<ConstraintViolation<Author>> violations = validator.validateProperty(noFirst, "firstName");
+    	
+    	//  then
+    	assertThat(violations.size()).isEqualTo(1);
+    }
+    
+    @Test
+    void notEmptyLastName() {
+    	// given
+    	Author noLast = new Author("Jostein", " ");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:150]: See previous comment about `noFirst`

**Review by knjk04 — CHANGES_REQUESTED** [review]
I've left some early feedback for you to address. Could you also please add tests for null (e.g. for a null author's first name)?

Once you've got everything working, could you also remove the annotations on the tag name in `Tag.java` and shelf name in `Shelf.java`

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 6]
```diff
@@ -114,4 +125,50 @@ void notRemoveNonOrphans() {
...
+    
+    @Test
+    void notAcceptBlankFirstName() {
+    	// given
+    	Author authorWithBlankFirstName = new Author(" ", "Gaarder");
+    	
+    	Set<ConstraintViolation<Author>> violations = validator.validateProperty(authorWithBlankFirstName, "firstName");
+ 
+    	//  then
+    	assertThat(violations.size()).isEqualTo(1);
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:159]: Instead of `isEqualTo(1)`, we can use the more concise `isOne()` method here

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 7]
```diff
@@ -114,4 +125,50 @@ void notRemoveNonOrphans() {
         // then
         assertThat(authorService.count()).isOne();
     }
+    
+    @Test
+    void notAcceptNullFirstName() {
+    	// given
+    	Author authorWithNullFirstName = new Author(null, "Gaarder");
+    	
+    	Set<ConstraintViolation<Author>> violations = validator.validateProperty(authorWithNullFirstName, "firstName");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:134]: As this exceeds our 100 character column limit, could you drop the initialisation part down one line?

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 8]
```diff
@@ -114,4 +125,50 @@ void notRemoveNonOrphans() {
...
+    	//  then
+    	assertThat(violations.size()).isEqualTo(2);
+    }
+    
+    @Test
+    void notAcceptNullLastName() {
+    	// given
+    	Author authorWithNullLastName = new Author("Jostein", null);
+    	
+    	Set<ConstraintViolation<Author>> violations = validator.validateProperty(authorWithNullLastName, "lastName");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:145]: exceeds 100 chars

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 9]
```diff
@@ -114,4 +125,50 @@ void notRemoveNonOrphans() {
...
+    	//  then
+    	assertThat(violations.size()).isEqualTo(2);
+    }
+    
+    @Test
+    void notAcceptBlankFirstName() {
+    	// given
+    	Author authorWithBlankFirstName = new Author(" ", "Gaarder");
+    	
+    	Set<ConstraintViolation<Author>> violations = validator.validateProperty(authorWithBlankFirstName, "firstName");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:156]: exceeds 100 chars

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 10]
```diff
@@ -114,4 +125,50 @@ void notRemoveNonOrphans() {
...
+    	//  then
+    	assertThat(violations.size()).isEqualTo(1);
+    }
+    
+    @Test
+    void notAcceptBlankLastName() {
+    	// given
+    	Author authorWithBlankLastName = new Author("Jostein", " ");
+    	
+    	Set<ConstraintViolation<Author>> violations = validator.validateProperty(authorWithBlankLastName, "lastName");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:167]: exceeds 100 chars

File: src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java [hunk 11]
```diff
@@ -114,4 +125,50 @@ void notRemoveNonOrphans() {
...
+   	
+    	//  then
+    	assertThat(violations.size()).isEqualTo(1);
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/AuthorTest.java:170]: isOne()

File: src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java [hunk 12]
```diff
@@ -28,6 +28,12 @@
 import org.springframework.transaction.annotation.Transactional;
 
 import java.util.Collections;
+import java.util.Set;
+
+import javax.validation.ConstraintViolation;
+import javax.validation.Validation;
+import javax.validation.Validator;
+import javax.validation.ValidatorFactory;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java:36]: I think we can remove this unused import

File: src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java [hunk 13]
```diff
@@ -157,4 +167,27 @@ void correctlyConvertNthEdition() {
...
+    	//  then
+    	assertThat(violations.size()).isEqualTo(2);
+    }
+    
+    @Test
+    void notAcceptBlankTitle() {
+    	// given
+    	Book bookWithBlankTitle = createBook(" ", predefinedShelfService.findToReadShelf());
+    	
+    	Set<ConstraintViolation<Book>> violations = validator.validateProperty(bookWithBlankTitle, "title");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java:188]: exceeds 100 chars

File: src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java [hunk 14]
```diff
@@ -157,4 +167,27 @@ void correctlyConvertNthEdition() {
...
         });
     }
+
+    
+    @Test
+    void notAcceptNullTitle() {
+    	// given
+    	Book bookWithNullTitle = createBook(null, predefinedShelfService.findToReadShelf());
+    	
+    	Set<ConstraintViolation<Book>> violations = validator.validateProperty(bookWithNullTitle, "title");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java:177]: exceeds 100 chars

File: src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java [hunk 15]
```diff
@@ -157,4 +167,27 @@ void correctlyConvertNthEdition() {
...
+    
+    @Test
+    void notAcceptBlankTitle() {
+    	// given
+    	Book bookWithBlankTitle = createBook(" ", predefinedShelfService.findToReadShelf());
+    	
+    	Set<ConstraintViolation<Book>> violations = validator.validateProperty(bookWithBlankTitle, "title");
+    	
+    	//  then
+    	assertThat(violations.size()).isEqualTo(1);
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/entity/BookTest.java:191]: isOne()

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for making the changes. I've left a few more suggestions. They're all style-related, so shouldn't take too long. 

Also, could you change your 'given' comments to 'when'?

**Review by knjk04 — APPROVED** [review]
Thanks for addressing the feedback, this is great!
